dot/network/discovery.go

```diff
@@ -0,0 +1,191 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	badger "github.com/ipfs/go-ds-badger2"
+	libp2phost "github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	libp2pdiscovery "github.com/libp2p/go-libp2p-discovery"
+	kaddht "github.com/libp2p/go-libp2p-kad-dht"
+	"github.com/libp2p/go-libp2p-kad-dht/dual"
+)
+
+// discovery handles discovery of new peers via the kademlia DHT
+type discovery struct {
+	ctx                context.Context
+	dht                *dual.DHT
+	h                  libp2phost.Host
+	bootnodes          []peer.AddrInfo
+	ds                 *badger.Datastore
+	pid                protocol.ID
+	minPeers, maxPeers int
+}
+
+func newDiscovery(ctx context.Context, h libp2phost.Host, bootnodes []peer.AddrInfo, ds *badger.Datastore, pid protocol.ID, min, max int) *discovery {
+	return &discovery{
+		ctx:       ctx,
+		h:         h,
+		bootnodes: bootnodes,
+		ds:        ds,
+		pid:       pid,
+		minPeers:  min,
+		maxPeers:  max,
+	}
+}
+
+// start creates the DHT.
+func (d *discovery) start() error {
+	if len(d.bootnodes) == 0 {
+		// get all currently connected peers and use them to bootstrap the DHT
+		peers := d.h.Network().Peers()
+
+		for {
+			if len(peers) == 0 {
+				logger.Info("no peers yet, waiting to start DHT...")
+				time.Sleep(time.Second * 10) // wait for peers to connect before starting DHT
+			} else {
+				break
+			}
+
+			peers = d.h.Network().Peers()
+		}
+
+		for _, p := range peers {
+			d.bootnodes = append(d.bootnodes, d.h.Peerstore().PeerInfo(p))
+		}
+	}
+
+	logger.Info("starting DHT...", "bootnodes", d.bootnodes)
+
+	dhtOpts := []dual.Option{
+		dual.DHTOption(kaddht.Datastore(d.ds)),
+		dual.DHTOption(kaddht.BootstrapPeers(d.bootnodes...)),
+		dual.DHTOption(kaddht.V1ProtocolOverride(d.pid + "/kad")),
+		dual.DHTOption(kaddht.Mode(kaddht.ModeAutoServer)),
+		//dual.DHTOption(kaddht.Mode(kaddht.ModeServer)),
+	}
+
+	// create DHT service
+	dht, err := dual.New(d.ctx, d.h, dhtOpts...)
+	if err != nil {
+		return err
+	}
+
+	d.dht = dht
+	return d.discoverAndAdvertise()
+}
+
+func (d *discovery) stop() error {
+	if d.dht == nil {
+		return nil
+	}
+
+	return d.dht.Close()
+}
+
+func (d *discovery) discoverAndAdvertise() error {
+	rd := libp2pdiscovery.NewRoutingDiscovery(d.dht)
+
+	err := d.dht.Bootstrap(d.ctx)
+	if err != nil {
+		return fmt.Errorf("failed to bootstrap DHT: %w", err)
+	}
+
+	// wait to connect to bootstrap peers
+	time.Sleep(time.Second)
+	peersToTry := make(map[*peer.AddrInfo]struct{})
```

Comment:
`peersToTry` could not be a slice of `*peer.AddrInfo`?

Reply:
it's a map so that there's no duplicates